File: core/src/main/java/dev/bee/kanjianki/core/HistoricalKanjiAggregate.java
package dev.bee.kanjianki.core;

public final class HistoricalKanjiAggregate {
    private final String kanji;
    private int activeCards;
    private int suspendedCards;
    private int matureSupportCount;
    private int totalLapses;
    private int totalReps;
    private int intervalCount;
    private double intervalSum;
    private int stabilityCount;
    private double stabilitySum;
    private int difficultyCount;
    private double difficultySum;
    private int retrievabilityCount;
    private double retrievabilitySum;
    private int weaknessScore;
    private String reasonCode = "";
    private int activeExampleCount;
    private int suspendedExampleCount;

    public HistoricalKanjiAggregate(String kanji) {
        this.kanji = kanji == null ? "" : kanji;
    }

    public void add(RecordsSyncModels.Card card, int matureDays) {
        addCard(
                card.intervalDays,
                card.reps,
                card.lapses,
                card.suspended,
                card.mature(matureDays),
                card.fsrsStability,
                card.fsrsDifficulty,
                card.fsrsRetrievability
        );
    }

    public void addCard(
            int intervalDays,
            int reps,
            int lapses,
            boolean suspended,
            boolean mature,
            Double fsrsStability,
            Double fsrsDifficulty,
            Double fsrsRetrievability
    ) {
        if (suspended) {
            suspendedCards++;
        } else {
            activeCards++;
        }
        if (mature) {
            matureSupportCount++;
        }
        totalLapses += Math.max(0, lapses);
        totalReps += Math.max(0, reps);
        intervalSum += Math.max(0, intervalDays);
        intervalCount++;
        if (fsrsStability != null) {
            stabilitySum += fsrsStability;
            stabilityCount++;
        }
        if (fsrsDifficulty != null) {
            difficultySum += fsrsDifficulty;
            difficultyCount++;
        }
        if (fsrsRetrievability != null) {
            retrievabilitySum += fsrsRetrievability;
            retrievabilityCount++;
        }
    }

    public void mergeDashboardEvidence(
            int rowWeaknessScore,
            String rowReasonCode,
            int rowActiveExampleCount,
            int rowSuspendedExampleCount,
            int rowMatureSupportCount
    ) {
        weaknessScore = rowWeaknessScore;
        reasonCode = rowReasonCode;
        activeExampleCount = Math.max(activeExampleCount, rowActiveExampleCount);
        suspendedExampleCount = Math.max(suspendedExampleCount, rowSuspendedExampleCount);
        matureSupportCount = Math.max(matureSupportCount, rowMatureSupportCount);
    }

    public String kanji() {
        return kanji;
    }

    public int activeCards() {
        return activeCards;
    }

    public int suspendedCards() {
        return suspendedCards;
    }

    public int matureSupportCount() {
        return matureSupportCount;
    }

    public int totalLapses() {
        return totalLapses;
    }

    public int totalReps() {
        return totalReps;
    }

    public double averageIntervalDays() {
        return intervalCount == 0 ? 0.0 : intervalSum / intervalCount;
    }

    public Double averageStability() {
        return stabilityCount == 0 ? null : stabilitySum / stabilityCount;
    }

    public Double averageDifficulty() {
        return difficultyCount == 0 ? null : difficultySum / difficultyCount;
    }

    public Double averageRetrievability() {
        return retrievabilityCount == 0 ? null : retrievabilitySum / retrievabilityCount;
    }

    public int weaknessScore() {
        return weaknessScore;
    }

    public String reasonCode() {
        return reasonCode;
    }

    public int activeExampleCount() {
        return activeExampleCount;
    }

    public int suspendedExampleCount() {
        return suspendedExampleCount;
    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/HistoricalKanjiAggregate.kt
package dev.bee.kanjianki.core

class HistoricalKanjiAggregate(kanji: String?) {
    private val kanji: String = kanji ?: ""
    private var activeCards = 0
    private var suspendedCards = 0
    private var matureSupportCount = 0
    private var totalLapses = 0
    private var totalReps = 0
    private var intervalCount = 0
    private var intervalSum = 0.0
    private var stabilityCount = 0
    private var stabilitySum = 0.0
    private var difficultyCount = 0
    private var difficultySum = 0.0
    private var retrievabilityCount = 0
    private var retrievabilitySum = 0.0
    private var weaknessScore = 0
    private var reasonCode = ""
    private var activeExampleCount = 0
    private var suspendedExampleCount = 0

    fun add(card: RecordsSyncModels.Card, matureDays: Int) {
        addCard(
            card.intervalDays,
            card.reps,
            card.lapses,
            card.suspended,
            card.mature(matureDays),
            FsrsMemoryValues(card.fsrsStability, card.fsrsDifficulty, card.fsrsRetrievability),
        )
    }

    fun addCard(
        intervalDays: Int,
        reps: Int,
        lapses: Int,
        suspended: Boolean,
        mature: Boolean,
        fsrs: FsrsMemoryValues?,
    ) {
        val memory = fsrs ?: EMPTY_FSRS
        if (suspended) {
            suspendedCards = saturatingAddNonNegative(suspendedCards, 1)
        } else {
            activeCards = saturatingAddNonNegative(activeCards, 1)
        }
        if (mature) {
            matureSupportCount = saturatingAddNonNegative(matureSupportCount, 1)
        }
        totalLapses = saturatingAddNonNegative(totalLapses, lapses)
        totalReps = saturatingAddNonNegative(totalReps, reps)
        intervalSum = saturatingFiniteAdd(intervalSum, maxOf(0, intervalDays).toDouble())
        intervalCount = saturatingAddNonNegative(intervalCount, 1)
        if (memory.stability?.isFinite() == true) {
            stabilitySum = saturatingFiniteAdd(stabilitySum, memory.stability)
            stabilityCount = saturatingAddNonNegative(stabilityCount, 1)
        }
        if (memory.difficulty?.isFinite() == true) {
            difficultySum = saturatingFiniteAdd(difficultySum, memory.difficulty)
            difficultyCount = saturatingAddNonNegative(difficultyCount, 1)
        }
        if (memory.retrievability?.isFinite() == true) {
            retrievabilitySum = saturatingFiniteAdd(retrievabilitySum, memory.retrievability)
            retrievabilityCount = saturatingAddNonNegative(retrievabilityCount, 1)
        }
    }

    fun mergeDashboardEvidence(
        rowWeaknessScore: Int,
        rowReasonCode: String?,
        rowActiveExampleCount: Int,
        rowSuspendedExampleCount: Int,
        rowMatureSupportCount: Int,
    ) {
        weaknessScore = rowWeaknessScore
        reasonCode = rowReasonCode ?: ""
        activeExampleCount = maxOf(activeExampleCount, rowActiveExampleCount)
        suspendedExampleCount = maxOf(suspendedExampleCount, rowSuspendedExampleCount)
        matureSupportCount = maxOf(matureSupportCount, rowMatureSupportCount)
    }

    fun kanji(): String = kanji

    fun activeCards(): Int = activeCards

    fun suspendedCards(): Int = suspendedCards

    fun matureSupportCount(): Int = matureSupportCount

    fun totalLapses(): Int = totalLapses

    fun totalReps(): Int = totalReps

    fun averageIntervalDays(): Double {
        return if (intervalCount == 0) 0.0 else intervalSum / intervalCount
    }

    fun averageStability(): Double? {
        return if (stabilityCount == 0) null else stabilitySum / stabilityCount
    }

    fun averageDifficulty(): Double? {
        return if (difficultyCount == 0) null else difficultySum / difficultyCount
    }

    fun averageRetrievability(): Double? {
        return if (retrievabilityCount == 0) null else retrievabilitySum / retrievabilityCount
    }

    fun weaknessScore(): Int = weaknessScore

    fun reasonCode(): String = reasonCode

    fun activeExampleCount(): Int = activeExampleCount

    fun suspendedExampleCount(): Int = suspendedExampleCount

    fun impactMetricSnapshot(): KanjiImpactAnalyzer.MetricSnapshot {
        return KanjiImpactAnalyzer.MetricSnapshot(
            activeCards,
            suspendedCards,
            matureSupportCount,
            averageIntervalDays(),
            totalReps,
            totalLapses,
            averageStability(),
            averageDifficulty(),
            averageRetrievability(),
        )
    }

    @JvmRecord
    data class FsrsMemoryValues(
        val stability: Double?,
        val difficulty: Double?,
        val retrievability: Double?,
    )

    companion object {
        private val EMPTY_FSRS = FsrsMemoryValues(null, null, null)

        private fun saturatingFiniteAdd(left: Double, right: Double): Double {
            val sum = left + right
            return when (sum) {
                Double.POSITIVE_INFINITY -> Double.MAX_VALUE
                Double.NEGATIVE_INFINITY -> -Double.MAX_VALUE
                else -> sum
            }
        }
    }
}
